loader/linux/icd_linux.c

```diff
@@ -36,111 +37,168 @@ static pthread_once_t initialized = PTHREAD_ONCE_INIT;
  *
  */
 
-// go through the list of vendors in the two configuration files
-void khrIcdOsVendorsEnumerate(void)
+typedef void khrIcdFileAdd(const char *);
+
+static inline void khrIcdOsDirEntryValidateAndAdd(const char *d_name, const char *path,
+                                                  const char *extension, khrIcdFileAdd addFunc)
+{
+    struct stat statBuff;
+    char* fileName = NULL;
+
+    // make sure the file name ends in `extension` (eg. .icd, or .lay)
+    if (strlen(extension) > strlen(d_name))
+    {
+        return;
+    }
+    if (strcmp(d_name + strlen(d_name) - strlen(extension), extension))
+    {
+        return;
+    }
+
+    // allocate space for the full path of the vendor library name
+    fileName = malloc(strlen(d_name) + strlen(path) + 2);
+    if (!fileName)
+    {
+        KHR_ICD_TRACE("Failed allocate space for ICD file path\n");
+        return;
+    }
+    sprintf(fileName, "%s/%s", path, d_name);
+
+    if (stat(fileName, &statBuff))
+    {
+        KHR_ICD_TRACE("Failed stat for: %s, continuing\n", fileName);
+        free(fileName);
+        return;
+    }
+
+    if (S_ISREG(statBuff.st_mode) || S_ISLNK(statBuff.st_mode))
+    {
+        FILE *fin = NULL;
+        char* buffer = NULL;
+        long bufferSize = 0;
+
+        // open the file and read its contents
+        fin = fopen(fileName, "r");
+        if (!fin)
+        {
+            free(fileName);
+            return;
+        }
+        fseek(fin, 0, SEEK_END);
+        bufferSize = ftell(fin);
+
+        buffer = malloc(bufferSize+1);
+        if (!buffer)
+        {
+            free(fileName);
+            fclose(fin);
+            return;
+        }
+        memset(buffer, 0, bufferSize+1);
+        fseek(fin, 0, SEEK_SET);
+        if (bufferSize != (long)fread(buffer, 1, bufferSize, fin))
+        {
+            free(fileName);
+            free(buffer);
+            fclose(fin);
+            return;
+        }
+        // ignore a newline at the end of the file
+        if (buffer[bufferSize-1] == '\n') buffer[bufferSize-1] = '\0';
+
+        // load the string read from the file
+        addFunc(buffer);
+
+        free(fileName);
+        free(buffer);
+        fclose(fin);
+     }
+     else
+     {
+         KHR_ICD_TRACE("File %s is not a regular file nor symbolic link, continuing\n", fileName);
+         free(fileName);
+     }
+}
+
+struct dirElem
+{
+    char *d_name;
+    unsigned char d_type;
+};
+
+static int compareDirElem(const void *a, const void *b)
+{
+    return strcoll(((struct dirElem *)a)->d_name, ((struct dirElem *)b)->d_name);
```

Comment:
I don't have a strong opinion about this but I do want to confirm that strcoll is intended here rather than strcmp, just because no other part of the ICD loader uses strcoll.  If it is intended, I'd suggest adding a comment to avoid confusion in the future.

Reply:
Yes it is intended, this is the way directory entries are sorted alphanumericaly on unixes (as far as I understand such things). I wanted our sorting to behave with minimum surprise for the user. I will add a comment.

Reply:
Digging deeper, maybe I should have used libc alphasort directly, didn't know about it...

Reply:
I see, thanks!  Adding a comment is fine for now - no need to make major changes.